lib/ansible/modules/cloud/amazon/ec2_ami.py

```diff
@@ -401,220 +392,289 @@ def create_image(module, ec2):
+        except botocore.exceptions.ClientError as e:
             if ('InvalidAMIID.NotFound' not in e.error_code and 'InvalidAMIID.Unavailable' not in e.error_code) and wait and i == wait_timeout - 1:
-                module.fail_json(msg="Error while trying to find the new image. Using wait=yes and/or a longer "
-                                     "wait_timeout may help. %s: %s" % (e.error_code, e.error_message))
+                module.fail_json(msg="Error while trying to find the new image. Using wait=yes and/or a longer wait_timeout may help. %s: %s"
+                                 % (e.error_code, e.error_message))
         finally:
             time.sleep(1)
 
-    if img.state != 'available':
-        module.fail_json(msg="Error while trying to find the new image. Using wait=yes and/or a longer wait_timeout may help.")
-
     if tags:
         try:
-            ec2.create_tags(image_id, tags)
-        except boto.exception.EC2ResponseError as e:
-            module.fail_json(msg="Image tagging failed => %s: %s" % (e.error_code, e.error_message))
+            connection.create_tags(Resources=[image_id], Tags=ansible_dict_to_boto3_tag_list(tags))
+        except botocore.exceptions.ClientError as e:
+            module.fail_json(msg="Error tagging image - " + str(e), exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
+
     if launch_permissions:
         try:
-            img = ec2.get_image(image_id)
-            img.set_launch_permissions(**launch_permissions)
-        except boto.exception.BotoServerError as e:
-            module.fail_json(msg="%s: %s" % (e.error_code, e.error_message), image_id=image_id)
-
-    module.exit_json(msg="AMI creation operation complete", changed=True, **get_ami_info(img))
+            image = get_image_by_id(module, connection, image_id)
+            image.set_launch_permissions(**launch_permissions)
+        except botocore.exceptions.ClientError as e:
+            module.fail_json(msg="Error setting launch permissions for image: " + image_id + " - " + str(e), exception=traceback.format_exc(),
+                             **camel_dict_to_snake_dict(e.response))
 
+    module.exit_json(msg="AMI creation operation complete.", changed=True, **get_ami_info(image))
 
-def deregister_image(module, ec2):
-    """
-    Deregisters AMI
-    """
 
+def deregister_image(module, connection):
     image_id = module.params.get('image_id')
     delete_snapshot = module.params.get('delete_snapshot')
     wait = module.params.get('wait')
-    wait_timeout = int(module.params.get('wait_timeout'))
+    wait_timeout = module.params.get('wait_timeout')
+    image = get_image_by_id(module, connection, image_id)
 
-    img = ec2.get_image(image_id)
-    if img is None:
-        module.fail_json(msg="Image %s does not exist" % image_id, changed=False)
+    if image is None:
+        module.fail_json(msg="Image %s does not exist." % image_id, changed=False)
```

Comment:
No! If the image doesn't exist, just exit with changed=False.

Reply:
I know this is the existing behaviour, but running the tests highlighted for me how wrong the existing behaviour is!

Reply:
Edit: I'm going to remove this objection, because it's existing behaviour, and suggest it as a fix in a separate PR - this one has been around the houses enough times!

Reply:
lol. no worries.